src/cuda/common.cu

```diff
@@ -199,6 +199,128 @@ void onedim_fseries_kernel_compute(CUFINUFFT_BIGINT nf, T *f, std::complex<doubl
+
+  return adjusted_ns * sizeof(cuda_complex<T>);
+}
+
+// Function to find bin_size_x == bin_size_y
+// where bin_size_x * bin_size_y * bin_size_z < mem_size
+// TODO: this can be done without a loop by using a direct formula
+template<typename T> int find_bin_size(std::size_t mem_size, int dim, int ns) {
+  int binsize = 1; // Start with the smallest possible bin size
+  while (true) {
+    // Calculate the shared memory required for the current bin_size_x and bin_size_y
+    std::size_t required_memory =
+        shared_memory_required<T>(dim, ns, binsize, binsize, binsize);
+
+    // Check if the required memory is less than the available memory
+    if (required_memory > mem_size) {
+      // If the condition is met, return the current bin_size_x
+      return binsize - 1;
+    }
+
+    // Increment bin_size_x for the next iteration
+    binsize++;
+  }
+}
+
+template<typename T>
+void cufinufft_setup_binsize(int type, int ns, int dim, cufinufft_opts *opts) {
+  // Marco Barbone 07/26/24. Using the shared memory available on the device, to
+  // determine the optimal binsize for the spreader.
+  // TODO: This can still be improved some sizes are hardcoded still
+  int shared_mem_per_block{}, device_id{};
+  switch (dim) {
+  case 1: {
+    if (opts->gpu_binsizex < 0) {
+      cudaGetDevice(&device_id);
+      if (const auto err = cudaGetLastError(); err != cudaSuccess) {
+        throw std::runtime_error(cudaGetErrorString(err));
+      }
+      cudaDeviceGetAttribute(&shared_mem_per_block,
+                             cudaDevAttrMaxSharedMemoryPerBlockOptin, device_id);
+      if (const auto err = cudaGetLastError(); err != cudaSuccess) {
+        throw std::runtime_error(cudaGetErrorString(err));
+      }
+      // use 1/6 of the shared memory for the binsize
+      // From experiments on multiple GPUs this gives the best tradeoff.
+      // It is within 90% of the maximum performance for all GPUs tested.
+      shared_mem_per_block /= 6;
+      const int bin_size =
+          shared_mem_per_block / sizeof(cuda_complex<T>) - ((ns + 1) / 2) * 2;
+      opts->gpu_binsizex = bin_size;
+    }
+    opts->gpu_binsizey = 1;
+    opts->gpu_binsizez = 1;
+  } break;
+  case 2: {
+    if (opts->gpu_binsizex < 0 || opts->gpu_binsizey < 0) {
+      switch (opts->gpu_method) {
+      case 0:
+      case 2: {
+        opts->gpu_binsizex = 32;
+        opts->gpu_binsizey = 32;
+      } break;
+      case 1: {
+        cudaGetDevice(&device_id);
+        if (const auto err = cudaGetLastError(); err != cudaSuccess) {
+          throw std::runtime_error(cudaGetErrorString(err));
+        }
+        cudaDeviceGetAttribute(&shared_mem_per_block,
+                               cudaDevAttrMaxSharedMemoryPerBlockOptin, device_id);
+        if (const auto err = cudaGetLastError(); err != cudaSuccess) {
+          throw std::runtime_error(cudaGetErrorString(err));
+        }
+
+        const auto binsize = find_bin_size<T>(shared_mem_per_block, dim, ns);
```

Comment:
Why not divide by 6 here?

Reply:
in 2D 1/6 is too small, it gets slower because of the excessive padding.

Reply:
Ok got it.